speed up SyncBatchNorm by batching distributed communication

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38246 speed up SyncBatchNorm by batching distributed communication**

Summary:

Speeds up SyncBatchNorm by batching the distributed communication.
Initial benchmarks show a ~15+% speed improvement on MobileNetV2 and
EfficientNetB3 on a single machine with 8 gpus. Improvement
vs baseline increases as # of gpus increases.

Test Plan:

```
python test/run_test.py -v -i distributed/test_distributed
# there were some test failures, but they were also present in master
```

verified that before+after intermediate values in fwd/bwd pass are equivalent (with `torch.allclose`)

benchmark runner:
https://gist.github.com/vkuzo/7b1ce1b1b051ee6d46877d0f18ab9b1f

results (1 forward pass + 1 backward pass, 1 machine, 8x Tesla-P100, batch_size=20 per node):
```
model           gpus  before_ms after_ms  speedup
efficientnet-b3 2     660       654       0.00909
efficientnet-b3 4     777       710       0.08623
efficientnet-b3 8     988       838       0.15182
mobilenet-v2    2     267       266       0.00375
mobilenet-v2    4     328       289       0.1189
mobilenet-v2    8     453       373       0.1766
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21505905](https://our.internmc.facebook.com/intern/diff/D21505905)

vkuzo committed:
speed up SyncBatchNorm by batching distributed communication

Summary:

Speeds up SyncBatchNorm by batching the distributed communication.
Initial benchmarks show a ~15% improvement on MobileNetV2 and
EfficientNetB3 on a single machine with 8 gpus. Improvement
vs baseline increases as # of gpus increases.

Test Plan:

benchmark runner:
https://gist.github.com/vkuzo/7b1ce1b1b051ee6d46877d0f18ab9b1f

results (1 machine, 8x Tesla-P100):
```
model           gpus  before_ms after_ms  speedup
efficientnet-b3 2     660       654       0.00909
efficientnet-b3 4     777       710       0.08623
efficientnet-b3 8     988       838       0.15182
mobilenet-v2    2     267       266       0.00375
mobilenet-v2    4     328       289       0.1189
mobilenet-v2    8     453       373       0.1766
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ngimel commented on torch/nn/modules/_functions.py:
```diff
@@ -15,23 +16,18 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
+        combined = torch.cat([mean, invstd, count], dim=0)
+        # world_size * (2C + 1)
+        combined_list = [
+            torch.zeros_like(combined) for k in range(world_size)
```
you don't need zeros_like here (it fills tensors with zeros) because values will be overwritten, right? torch.empty_like is faster and should be enough.

vkuzo committed:
Update on "speed up SyncBatchNorm by batching distributed communication"


Summary:

Speeds up SyncBatchNorm by batching the distributed communication.
Initial benchmarks show a ~15+% speed improvement on MobileNetV2 and
EfficientNetB3 on a single machine with 8 gpus. Improvement
vs baseline increases as # of gpus increases.

Test Plan:

```
python test/run_test.py -v -i distributed/test_distributed
# there were some test failures, but they were also present in master
```

verified that before+after intermediate values in fwd/bwd pass are equivalent (with `torch.allclose`)

benchmark runner:
https://gist.github.com/vkuzo/7b1ce1b1b051ee6d46877d0f18ab9b1f

results (1 forward pass + 1 backward pass, 1 machine, 8x Tesla-P100, batch_size=20 per node):
```
model           gpus  before_ms after_ms  speedup
efficientnet-b3 2     660       654       0.00909
efficientnet-b3 4     777       710       0.08623
efficientnet-b3 8     988       838       0.15182
mobilenet-v2    2     267       266       0.00375
mobilenet-v2    4     328       289       0.1189
mobilenet-v2    8     453       373       0.1766
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21505905](https://our.internmc.facebook.com/intern/diff/D21505905)

[ghstack-poisoned]

raghuramank100 commented on torch/nn/modules/_functions.py:
```diff
@@ -15,23 +16,18 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
+            torch.empty_like(combined) for k in range(world_size)
+        ]
+        # Use allgather instead of allreduce since I don't trust in-place operations ..
+        dist.all_gather(combined_list, combined, async_op=False)
```
For my understanding, why is the async_op argument changes from True to False here?

vkuzo replied:
this is because now there is only one all_gather call, so there is no difference between calling it async and waiting versus calling it sync, since it needs to finish before the next block of code.  Before this diff, there were three all_gather calls, all launched asynchronously.

pritamdamania87 commented on torch/nn/modules/_functions.py:
```diff
@@ -15,23 +16,18 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
+        combined_list = [
+            torch.empty_like(combined) for k in range(world_size)
+        ]
+        # Use allgather instead of allreduce since I don't trust in-place operations ..
```
@vkuzo Wondering if there is a better reason than this not to use allreduce :) We use in-place allreduce operations a lot and in general its much faster than allgather.

ngimel replied:
You'd need to do a weighted sum here and not just all-reduce. Here the amount of data transfered is very small, so you'd be limited just by the latency of operation which should be approximately the same for allreduce and allgather.

vkuzo replied:
I'm definitely not an expert in this - after finding that this was a bottleneck, for this diff I took the approach of minimizing risk and copying the battle tested piece of the corresponding `detectron2` layer (including this comment :) ).

do we have an estimate of performance savings of using allreduce? We can update it if it's worth it.

pritamdamania87 replied:
@ngimel Was looking through this code again and wasn't sure about the weighted all-reduce part. Don't we need to find the global mean and variance for SyncBN and that can be done using all-reduce?
